mammoth/model_builder.py

```diff
@@ -190,25 +152,89 @@ def build_xcoder(
                 )
             else:
                 raise ValueError(f'Unrecognized adapter_type {adapter_opts["adapter_type"]}')
+            layer_stack_index = adapter_params['layer_stack_index']
             adapter = Adapter(
                 adapter_group=component.adapter_group,
                 sub_id=component.sub_id,
+                layer_stack_index=layer_stack_index,
             )
             adapters_by_name[adapter.name] = adapter
             for layer_idx in adapter_params['layers']:
                 adapter_layer = adapter_layer_func()
                 adapter.add_layer(layer_idx, adapter_layer)
-            layer_stack_index = adapter_params['layer_stack_index']
-            for xcoder_id, attention_layers in attention_layer_blocks[layer_stack_index].items():
+    else:
+        adapters_by_name = None
+    return adapters_by_name
+
+
+def build_xcoder(
+    side: Side,
+    model_opts,
+    vocabs_dict: Dict[Tuple[str, str], Vocab],
+    device,
+    task_queue_manager,
+    single_task: Optional[str] = None,
+    token_embs: Optional[Dict[str, Vocab]] = None,
+    adapters_by_name: Optional[Dict[str, Adapter]] = None,
+) -> StackXcoder:
+    """
+    Build a StackXcoder for use as either Encoder or Decoder.
+    side: a Side enum from distributed components
+    model_opts: options
+    vocabs_dict: A dict mapping ('src'|'tgt', lang) to a Vocab.
+    device: torch.device
+    task_queue_manager: TaskQueueManager
+    single_task: if a task_id string is given, the built model contains only the components necessary for that task.
+    token_embs: to tie encoder and decoder embeddings, pass existing embeddings here.
+    """
+    my_components: List[DistributedComponent] = task_queue_manager.get_my_distributed_components()
+    my_components = [
```

Comment:
not fond of redefining the var immediately, either rename (e.g. "side specific components"), or merge with 190?

Reply:
Now there are `my_components`, `my_side_specific_components`, and `components_to_create`. Not sure if better, but at least more verbose.